libraries/stdlib/jvm/src/kotlin/text/StringNumberConversionsJVM.kt

```diff
@@ -239,35 +239,215 @@ public fun String.toBigDecimalOrNull(): java.math.BigDecimal? =
-                "(($HexString)[pP][+-]?$Digits)"            // HexSignificand BinaryExponent
-
-        val fpRegex = "[\\x00-\\x20]*[+-]?(NaN|Infinity|(($Number)[fFdD]?))[\\x00-\\x20]*"
-
-        Regex(fpRegex)
-    }
-}
-
 private inline fun <T> screenFloatValue(str: String, parse: (String) -> T): T? {
     return try {
-        if (ScreenFloatValueRegEx.value.matches(str))
+        if (isValidFloat(str))
             parse(str)
         else
             null
-    } catch (e: NumberFormatException) {  // overflow
+    } catch (_: NumberFormatException) {  // overflow
         null
     }
 }
+
+private const val LengthOfNaN = 2 // "NaN".length - 1
+private const val LengthOfInfinity = 7 // "Infinity".length - 1
+
+private fun isValidFloat(s: String): Boolean {
+    // A float can have one of two representations:
+    //
+    // 1. Standard:
+    //     - With an integer part only: 1234
+    //     - With integer and fractional parts: 1234.4678
+    //     - With a fractional part only: .4678
+    //
+    //     Optional sign prefix: + or -
+    //     Optional signed exponent: e or E, followed by optionally signed digits (+12, -12, 12)
+    //     Optional suffix: f, F, d, or D (for instance 12.34f or .34D)
+    //
+    // 2. Hexadecimal:
+    //     - With an integer part only: 0x12ab
+    //     - With integer and fractional parts: 0x12ab.CD78
+    //     - With a fractional part only: 0x.CD78
+    //
+    //     Mandatory signed exponent: p or P, followed by optionally signed digits (+12, -12, 12)
+    //
+    //     Optional sign prefix: + or -
+    //     Optional suffix: f, F, d, or D (for instance 12.34f or .34D)
+    //
+    // Two special cases:
+    //     "NaN" and "Infinity" strings, can have an optional sign prefix (+ or -)
+    //
+    // Implementation notes:
+    //     - The pattern "myChar.code or 0x20 == 'x'.code" is used to perform a case-insensitive
+    //       comparison of a character. Adding the 0x20 bit turns an upper case ASCII letter into
+    //       a lower case one.
+
+    var start = 0
+    var end = s.length - 1
+
+    // Skip leading spaces
+    while (start <= end && s[start].code <= 0x20) start++
+
+    // Empty/whitespace string
+    if (start > end) return false
+
+    // Skip trailing spaces
+    while (end > start && s[end].code <= 0x20) end--
+
+    // Number starts with a positive or negative sign
+    if (s[start] == '+' || s[start] == '-') start++
+    // If we have nothing after the sign, the string is invalid
+    if (start > end) return false
+
+    var hasIntegerPart: Boolean
+    var hasFractionalPart = false
+    var isHex = false
+
+    // Might be a hex string
+    if (s[start] == '0') {
+        start++
+        // A "0" on its own is valid
+        if (start > end) return true
+
+        // Test for [xX] to see if we truly have a hex string
+        if (s[start].code or 0x20 == 'x'.code) {
+            start++
+
+            // Look for hex digits after the 0x prefix
```

Comment:
Mantissa validation is almost identical for both hex and decimal formats. The only major differences are:
- a digit-validating predicate;
- integral part is mandatory for hexadecimal literals.

It could be outlined to a function like `internal fun String.validateMantissaAndAdvance(start: Int, endInclusive: Int, intPartRequired: Boolean, predicate: (Char) -> Boolean): Int`.
Validation failure could be represented by good old `-1` returned from that function (and the effect on the overall performance seems to be neglectable).

Reply:
The non-hex case also needs to check for Infinity/NaN. I'll do what you suggest but I find this kind of code reuse not that useful as you now have a boolean affecting a behavior that's also linked to the predicate. It technically makes the call site a bit simpler, but you do introduce complexity of understanding how that function behaves in both cases.